central/complianceoperator/v2/profiles/datastore/datastore_impl.go

```diff
@@ -81,3 +85,115 @@ func (d *datastoreImpl) GetProfilesByClusters(ctx context.Context, clusterIDs []
+	readableClusterIDs := bestEffortClusters(ctx, clusterIDs)
+
+	var err error
+	q, err = withSACFilter(ctx, resources.Compliance, q)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedQuery := search.NewQueryBuilder().
+		AddExactMatches(search.ClusterID, readableClusterIDs...).
+		AddNumericField(search.ProfileCount, storage.Comparator_EQUALS, float32(len(readableClusterIDs))).
+		ProtoQuery()
+
+	parsedQuery = search.ConjunctionQuery(
+		parsedQuery,
+		q,
+	)
+
+	// Build the select and group by on distinct profile name
+	parsedQuery.Selects = []*v1.QuerySelect{
+		search.NewQuerySelect(search.ComplianceOperatorProfileName).Distinct().Proto(),
+	}
+	parsedQuery.GroupBy = &v1.QueryGroupBy{
+		Fields: []string{
+			search.ComplianceOperatorProfileName.String(),
+		},
+	}
+	parsedQuery.Pagination = q.GetPagination()
+
+	var results []*distinctProfileName
+	results, err = pgSearch.RunSelectRequestForSchema[distinctProfileName](ctx, d.db, schema.ComplianceOperatorProfileV2Schema, parsedQuery)
+	if err != nil {
+		return nil, err
+	}
+	if len(results) == 0 {
+		return nil, nil
+	}
+	profileNames := make([]string, 0, len(results))
+	for _, result := range results {
+		profileNames = append(profileNames, result.ProfileName)
+	}
+
+	return profileNames, err
+}
+
+type distinctProfileCount struct {
+	TotalCount int    `db:"compliance_profile_name_count"`
+	Name       string `db:"compliance_profile_name"`
+}
+
+// CountDistinctProfiles returns count of distinct profiles matching query
+func (d *datastoreImpl) CountDistinctProfiles(ctx context.Context, q *v1.Query, clusterIDs []string) (int, error) {
+	// Build the matching query to restrict profiles to the incoming clusters
+	readableClusterIDs := bestEffortClusters(ctx, clusterIDs)
+
+	query := search.ConjunctionQuery(
+		search.NewQueryBuilder().
+			AddExactMatches(search.ClusterID, readableClusterIDs...).
+			AddNumericField(search.ProfileCount, storage.Comparator_EQUALS, float32(len(readableClusterIDs))).ProtoQuery(),
+		q,
+	)
+
+	query.GroupBy = &v1.QueryGroupBy{
+		Fields: []string{
+			search.ComplianceOperatorProfileName.String(),
+		},
+	}
+
+	var results []*distinctProfileCount
+	results, err := pgSearch.RunSelectRequestForSchema[distinctProfileCount](ctx, d.db, schema.ComplianceOperatorProfileV2Schema, withCountQuery(query, search.ComplianceOperatorProfileName))
+	if err != nil {
+		return 0, err
+	}
+	return len(results), nil
+}
+
+func withCountQuery(q *v1.Query, field search.FieldLabel) *v1.Query {
+	cloned := q.Clone()
+	cloned.Selects = []*v1.QuerySelect{
+		search.NewQuerySelect(field).AggrFunc(aggregatefunc.Count).Proto(),
+	}
+	return cloned
+}
+
+func bestEffortClusters(ctx context.Context, clusterIDs []string) []string {
+	// Best effort SAC.  We only want to return profiles from the cluster list that the user has access to
+	// view.  So we do the access check to build a narrowed list vs embedding in the query as the logic is far simpler.
```

Comment:
```suggestion
	// view.  So we perform an access check to create a narrowed list instead of embedding it in the query, as the logic is much simpler.
```